### PR TITLE
Adding cte_follow_insert and test

### LINKED ISSUE #380: Generating Insert Querys with CTEs put them at the top of the whole statement incorrectly
when you build a cte for a select statement used as the target of an insert from_select, it renders the with statement before the insert statement which athena complains about with a `mismatched input 'INSERT'. Expecting: '(', 'SELECT', 'TABLE', 'VALUES'` error.  This is how it was fixed for MySQL and oracle dialects a while back https://github.com/sqlalchemy/sqlalchemy/commit/3619edcb8aa3ceef2a44925b85315fc0e90c5982 @zzzeek

### PR DESCRIPTION
**Pull request by jasonamyers**
This should help fix #380 I don't have all the test infrastructure, but this shouldn't relay on anything other than the dialect.
Signed-off-by: Jason Myers <jason.myers@stratasan.com>

### CODE REVIEW

File: tests/test_sqlalchemy_athena.py [hunk 1]
```diff
@@ -1331,3 +1331,35 @@ def test_create_table_with_date_partition(self, engine):
...
+
+
+        cte = select([my_table.c.name]).where(my_table.c.name == 'bar').cte()
+
+        sel = select([my_table.c.myid, my_table.c.name]).where(
+            my_table.c.name == cte.c.name)
+
+        ins = my_other_table.insert().\
+            from_select(("otherid", "othername"), sel)
+        self.assert_compile(
```
Comment (laughingman7743) [inline tests/test_sqlalchemy_athena.py:1355]: The method assert_compile seems to be specific to the SQLAlchemy test class, so it needs to be changed to use assert, which Python supports by default.

**Comment by laughingman7743** [comment]
In addition to compiling the statement, you will also need to test it so that you can actually execute the query to Athena and check the results.
https://github.com/sqlalchemy/sqlalchemy/commit/3619edcb8aa3ceef2a44925b85315fc0e90c5982#diff-e1df06623173c42f17127789e3096b698d3da0c655207f30e42909ffd15c05d3R11-R193
I don't think we need the same tests as SQLAlchemy. A simpler test would be fine.

**Comment by jasonamyers** [comment]
> In addition to compiling the statement, you will also need to test it so that you can actually execute the query to Athena and check the results. [sqlalchemy/sqlalchemy@3619edc#diff-e1df06623173c42f17127789e3096b698d3da0c655207f30e42909ffd15c05d3R11-R193](https://github.com/sqlalchemy/sqlalchemy/commit/3619edcb8aa3ceef2a44925b85315fc0e90c5982#diff-e1df06623173c42f17127789e3096b698d3da0c655207f30e42909ffd15c05d3R11-R193) I don't think we need the same tests as SQLAlchemy. A simpler test would be fine.

I'll work on this Friday of this week.

**Comment by laughingman7743** [comment]
```
    def test_insert_from_select_cte_follows_insert_one(self, engine):
        engine, conn = engine
        metadata = MetaData(schema=ENV.schema)
        table_name = "select_cte_insert_one_1"
        table = Table(
            table_name,
            metadata,
            Column("id", types.Integer),
            Column("name", types.String(30)),
            Column("description", types.String(30)),
        )
        other_table_name = "select_cte_insert_one_2"
        other_table = Table(
            other_table_name,
            metadata,
            Column("id", types.Integer, primary_key=True),
            Column("name", types.String(30)),
        )

        cte = sqlalchemy.select([table.c.name]).where(table.c.name == "bar").cte()
        sel = sqlalchemy.select([table.c.id, table.c.name]).where(table.c.name == cte.c.name)
        ins = other_table.insert().from_select(("id", "name"), sel)

        table.create(bind=conn)
        other_table.create(bind=conn)
        conn.execute(
            table.insert(),
            [
                {"id": 1, "name": "foo", "description": "description foo"},
                {"id": 2, "name": "bar", "description": "description bar"},
            ],
        )
        conn.execute(ins)
        actual = conn.execute(sqlalchemy.select([other_table])).fetchall()

        assert (
            str(ins)
            == textwrap.dedent(
                f"""
            WITH anon_1 AS \n\
            (SELECT "{ENV.schema}".{table_name}.name AS name \n\
            FROM "{ENV.schema}".{table_name} \n\
            WHERE "{ENV.schema}".{table_name}.name = :name_1)
             INSERT INTO "{ENV.schema}".{other_table_name} (id, name) \
SELECT "{ENV.schema}".{table_name}.id, "{ENV.schema}".{table_name}.name \n\
            FROM "{ENV.schema}".{table_name}, anon_1 \n\
            WHERE "{ENV.schema}".{table_name}.name = anon_1.name \n\
            """
            ).strip()
        )
        assert actual == [(2, "bar")]
```
I have tried to run such a test but get an error, Athena does not seem to support such a query, what do you think? Could you please share a simple example where such a query would succeed?
````
self = <pyathena.cursor.Cursor object at 0x11400dc60>
operation = 'WITH anon_1 AS \n(SELECT default.select_cte_insert_one_1.name AS name \nFROM default.select_cte_insert_one_1 \nWHERE ...t_one_1.name \nFROM default.select_cte_insert_one_1, anon_1 \nWHERE default.select_cte_insert_one_1.name = anon_1.name'
parameters = {'name_1': 'bar'}, work_group = None, s3_staging_dir = None
cache_size = 0, cache_expiration_time = 0

    def _execute(
        self,
        operation: str,
        parameters: Optional[Dict[str, Any]] = None,
        work_group: Optional[str] = None,
        s3_staging_dir: Optional[str] = None,
        cache_size: int = 0,
        cache_expiration_time: int = 0,
    ) -> str:
        query = self._formatter.format(operation, parameters)
        _logger.debug(query)
    
        request = self._build_start_query_execution_request(query, work_group, s3_staging_dir)
        query_id = self._find_previous_query_id(
            query,
            work_group,
            cache_size=cache_size,
            cache_expiration_time=cache_expiration_time,
        )
        if query_id is None:
            try:
                query_id = retry_api_call(
                    self._connection.client.start_query_execution,
                    config=self._retry_config,
                    logger=_logger,
                    **request,
                ).get("QueryExecutionId", None)
            except Exception as e:
                _logger.exception("Failed to execute query.")
>               raise DatabaseError(*e.args) from e
E               sqlalchemy.exc.DatabaseError: (pyathena.error.DatabaseError) An error occurred (InvalidRequestException) when calling the StartQueryExecution operation: line 5:2: mismatched input 'INSERT'. Expecting: '(', 'SELECT', 'TABLE', 'VALUES'
E               [SQL: WITH anon_1 AS 
E               (SELECT default.select_cte_insert_one_1.name AS name 
E               FROM default.select_cte_insert_one_1 
E               WHERE default.select_cte_insert_one_1.name = %(name_1)s)
E                INSERT INTO default.select_cte_insert_one_2 (id, name) SELECT default.select_cte_insert_one_1.id, default.select_cte_insert_one_1.name 
E               FROM default.select_cte_insert_one_1, anon_1 
E               WHERE default.select_cte_insert_one_1.name = anon_1.name]
E               [parameters: {'name_1': 'bar'}]
E               (Background on this error at: https://sqlalche.me/e/14/4xp6)

../../pyathena/common.py:502: DatabaseError
```
Also, you must be able to run the tests yourself. Please set it up yourself in the forked repository.

**Comment by jasonamyers** [comment]
Your test is failing because it is generating the WITH statement before the INSERT statement. What branch are you adding this test too, and does it have the `cte_follows_insert = True` set to true on the dialect?

**Comment by laughingman7743** [comment]
Ah, I forgot to enable cte_follows_insert. The query seems to succeed.

**Comment by jasonamyers** [comment]
> Ah, I forgot to enable cte_follows_insert. The query seems to succeed.

Would you like me to copy this into my PR? if so is this the missing tests you'd want?

**Comment by jasonamyers** [comment]
> I am trying the following branch. https://github.com/laughingman7743/PyAthena/pull/384/files You may copy these changes to this branch.

I'm kinda surprised that 
``` 
str(ins)
            == textwrap.dedent(
                f"""
            WITH anon_1 AS \n\
            (SELECT "{ENV.schema}".{table_name}.name AS name \n\
            FROM "{ENV.schema}".{table_name} \n\
            WHERE "{ENV.schema}".{table_name}.name = :name_1)
             INSERT INTO "{ENV.schema}".{other_table_name} (id, name) \
SELECT "{ENV.schema}".{table_name}.id, "{ENV.schema}".{table_name}.name \n\
            FROM "{ENV.schema}".{table_name}, anon_1 \n\
            WHERE "{ENV.schema}".{table_name}.name = anon_1.name \n\
            """
```
 is passing. I'd expect that to need to be 
```
 str(ins)
            == textwrap.dedent(
                f"""
             INSERT INTO "{ENV.schema}".{other_table_name} (id, name) \
WITH anon_1 AS \n\
            (SELECT "{ENV.schema}".{table_name}.name AS name \n\
            FROM "{ENV.schema}".{table_name} \n\
            WHERE "{ENV.schema}".{table_name}.name = :name_1)
SELECT "{ENV.schema}".{table_name}.id, "{ENV.schema}".{table_name}.name \n\
            FROM "{ENV.schema}".{table_name}, anon_1 \n\
            WHERE "{ENV.schema}".{table_name}.name = anon_1.name \n\
            """
```

**Comment by jasonamyers** [comment]
Pushed an update to my branch with that test adjusted a bit.

**Comment by laughingman7743** [comment]
https://github.com/laughingman7743/PyAthena/pull/384

**Comment by jasonamyers** [comment]
@laughingman7743 thank you so much for the fix!

**Pull request closed without merging**